fetch.bs

```diff
@@ -2697,15 +2700,87 @@ functionality.
-<a for="fetch record">controller</a>.
+<p>When a <a for=fetch>fetch group</a> <var>fetchGroup</var> is
+<dfn export for="fetch group" id=concept-fetch-group-terminate>terminated</dfn>:
+
+<ol>
+ <li><p>For each associated <a for="fetch group">fetch record</a> <var>record</var>,
+ if <var>record</var>'s <a for="fetch record">controller</a> is non-null and
+ <var>record</var>'s <a for="fetch record">request</a>'s <a>done flag</a> is unset and
+ <a for=request>keepalive</a> is false, <a for="fetch controller">terminate</a> <var>record</var>'s
+ <a for="fetch record">controller</a>.
+
+ <li><p><a for=list>For each</a> <a for="fetch group">deferred fetch record</a>
+ <var>deferredRecord</var> in <var>fetchGroup</var>'s
+ <a for="fetch group">deferred fetch records</a>: If the result of atomically exchanging the value
+ of <var>deferredRecord</var>'s <a for="deferred fetch record">invoke state</a> with
+ "<code>terminated</code>" is not "<code>sent</code>", then the user agent should
+ <a for=/>fetch</a> <var>deferredRecord</var>'s <a for="deferred fetch record">request</a>. The
+ exact time in which the <a for=/>fetch</a> takes place is <a>implementation-defined</a>.
+</ol>
+
+<p>When a <a for=fetch>fetch group</a> <var>fetchGroup</var> is
+<dfn export for="fetch group" id=concept-fetch-group-deactivate>deactivated</dfn>:
+<a for=list>For each</a> <a>deferred fetch record</a> <var>deferredRecord</var> in
+<var>fetchGroup</var>'s <a for="fetch group">deferred fetch records</a> whose
+<a for="deferred fetch record">inactivity deferred delay</a> is not null:
+
+<ol>
+ <li><p>Set <var>deferredRecord</var>'s <a for="deferred fetch record">invoke state</a> to
+ "<code>scheduled</code>".
+
+ <li>
+  <p>Run the following steps <a>in parallel</a>:
+
+  <ol>
+   <li><p>The user agent should wait until <var>deferredRecord</var>'s
+   <a for="deferred fetch record">inactivity deferred delay</a> have passed or until
+   <var>deferredRecord</var>'s <a for="deferred fetch record">invoke state</a> is not
+   "<code>scheduled</code>". The user agent may wait for a longer or shorter period time, e.g., to
+   optimize batching of deferred fetches.
+
+   <li><p>If the result of atomically exchanging the value of <var>deferredRecord</var>'s
+   <a for="deferred fetch record">invoke state</a> with "<code>sent</code>" is
+   "<code>scheduled</code>", then <a for=/>fetch</a> <var>record</var>'s
+   <a for="fetch record">request</a>.
+  </ol>
+ </li>
+</ol>
+
+<p>When a <a for=fetch>fetch group</a> <var>fetchGroup</var> is
+<dfn export for="fetch group" id=concept-fetch-group-reactivate>reactivated</dfn>:
+<a for=list>For each</a> <a for=/>deferred fetch record</a> <var>deferredRecord</var> in
+<var>fetchGroup</var>'s <a for="fetch group">deferred fetch records</a>: If the result of atomically
+exchanging the value of <var>deferredRecord</var>'s <a for="deferred fetch record">invoke state</a>
+with "<code>deferred</code>" is "<code>sent</code>", then <a for=list>remove</a>
```

Comment:
> <p>When a <a for=fetch>fetch group</a> <var>fetchGroup</var> is <dfn export for="fetch group" id=concept-fetch-group-reactivate>reactivated</dfn>:
> If the result of atomically exchanging the value of <var>deferredRecord</var>'s <a for="deferred fetch record">invoke state</a> with "<code>deferred</code>" is "<code>sent</code>", then <a for=list>remove</a> <var>deferredRecord</var> from <var>fetchGroup</var>'s <a for="fetch group">deferred fetch records</a> and set <var>deferredRecord</var>'s <a for="deferred fetch record">sent</a> to true.

Assuming `reactivated` means going out of bfcache (background to foreground): I feel that `sent` (or `committed`) deferredRecord should not be remove here. Can't this also be done in `deactivated` just following after the step that performs fetching the request?

Reply:
To add more, I think this API's behavior is `send (commit) & forget`. If a document may be kept in bfcache for an uncertain amount of time thus may not become `reactivated`. If a deferredRecord is already sent when deactivated, isn't it better to just drop it?

Reply:
Doing it anywhere else would be racy. Apart from `invokeState`, the list of `deferred fetch records` should not be accessed in parallel, as in, should only be accessed from the main thread.
If we did this in `deactivated`, there would be a short window where the fetch record is still in the list even though it's committed, which means it would, for example, affect the deferred fetch quota.

Do you see any problem with doing this here?